File: csharp/Output.cs
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CSharpBenchmark;

public class SystemInfo
{
    [JsonPropertyName("os")]
    public string Os { get; set; } = string.Empty;
    
    [JsonPropertyName("arch")]
    public string Arch { get; set; } = string.Empty;
    
    [JsonPropertyName("cpus")]
    public int Cpus { get; set; }
}

public class BenchmarkReport
{
    [JsonPropertyName("language")]
    public string Language { get; set; } = "csharp";
    
    [JsonPropertyName("timestamp")]
    public string Timestamp { get; set; } = string.Empty;
    
    [JsonPropertyName("dotnet_version")]
    public string DotnetVersion { get; set; } = string.Empty;
    
    [JsonPropertyName("runtime_version")]
    public string RuntimeVersion { get; set; } = string.Empty;
    
    [JsonPropertyName("system")]
    public SystemInfo System { get; set; } = new();
    
    [JsonPropertyName("tests")]
    public List<BenchmarkResult> Tests { get; set; } = new();
    
    [JsonPropertyName("total_time_seconds")]
    public double TotalTimeSeconds { get; set; }
}

public static class Output
{
    public static void PrintResults(List<BenchmarkResult> results)
    {
        Console.WriteLine("\n=== BENCHMARK RESULTS ===");
        foreach (var result in results)
        {
            Console.WriteLine($"Test: {result.Test}");
            Console.WriteLine($"  Duration: {result.DurationNs} ns");
            Console.WriteLine($"  Memory: {result.MemoryBytes} bytes");
            Console.WriteLine($"  Operations: {result.Operations}");
            Console.WriteLine($"  Ops/sec: {result.OpsPerSec:F2}");
            Console.WriteLine();
        }
    }
    
    public static void SaveResultsToJson(List<BenchmarkResult> results, double totalTime)
    {
        var report = new BenchmarkReport
        {
            Timestamp = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss"),
            DotnetVersion = Environment.Version.ToString(),
            RuntimeVersion = RuntimeInformation.FrameworkDescription,
            System = new SystemInfo
            {
                Os = GetOsName(),
                Arch = RuntimeInformation.OSArchitecture.ToString(),
                Cpus = Environment.ProcessorCount
            },
            Tests = results,
            TotalTimeSeconds = totalTime
        };
        
        var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
        var filename = $"benchmark_csharp_{timestamp}.json";
        
        try
        {
            var options = new JsonSerializerOptions
            {
                WriteIndented = true
            };
            
            var json = JsonSerializer.Serialize(report, options);
            File.WriteAllText(filename, json);
            Console.WriteLine($"Results saved to: {filename}");
        }
        catch (Exception e)
        {
            throw new InvalidOperationException($"Error writing file: {e.Message}", e);
        }
    }
    
    private static string GetOsName()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            return "Windows";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            return "Linux";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            return "macOS";
        return "Unknown";
    }
}

File: typescript/src/output.ts
import * as fs from 'fs';
import * as os from 'os';
import { BenchmarkResult } from './benchmark';

interface BenchmarkReport {
    language: string;
    timestamp: string;
    node_version: string;
    typescript_version?: string;
    system: {
        os: string;
        arch: string;
        cpus: number;
    };
    tests: BenchmarkResult[];
    total_time_seconds: number;
}

export function printResults(results: BenchmarkResult[]): void {
    console.log("\n=== BENCHMARK RESULTS ===");
    for (const result of results) {
        console.log(`Test: ${result.test}`);
        console.log(`  Duration: ${result.duration_ns} ns`);
        console.log(`  Memory: ${result.memory_bytes} bytes`);
        console.log(`  Operations: ${result.operations}`);
        console.log(`  Ops/sec: ${result.ops_per_sec.toFixed(2)}`);
        console.log();
    }
}

export function saveResultsToJSON(results: BenchmarkResult[], totalTime: number): void {
    const report: BenchmarkReport = {
        language: 'typescript',                          // 言語名
        timestamp: new Date().toISOString(),             // 実行時刻
        node_version: process.version,                   // Node.jsのバージョン
        typescript_version: require('typescript/package.json').version, // TypeScriptのバージョン
        system: {
            os: os.platform(),                          // OS名
            arch: os.arch(),                            // アーキテクチャ
            cpus: os.cpus().length                      // CPU数
        },
        tests: results,                                 // ベンチマーク結果
        total_time_seconds: totalTime                   // 総実行時間(秒)
    };
    
    const timestamp = new Date().toISOString()
        .replace(/[:.]/g, '')
        .replace('T', '_')
        .substring(0, 15);
    
    const filename = `benchmark_typescript_${timestamp}.json`;
    
    try {
        fs.writeFileSync(filename, JSON.stringify(report, null, 2));
        console.log(`Results saved to: ${filename}`);
    } catch (error) {
        throw new Error(`Error writing file: ${(error as Error).message}`);
    }
}